lib/bolt/transport/docker/connection.rb

```diff
@@ -29,203 +48,113 @@ def connect
           true
         rescue StandardError => e
           raise Bolt::Node::ConnectError.new(
-            "Failed to connect to #{@target.safe_name}: #{e.message}",
+            "Failed to connect to #{target.safe_name}: #{e.message}",
             'CONNECT_ERROR'
           )
         end
 
-        # Executes a command inside the target container
+        # Executes a command inside the target container. This is called from the shell class.
         #
-        # @param command [Array] The command to run, expressed as an array of strings
-        # @param options [Hash] command specific options
-        # @option opts [String] :interpreter statements that are prefixed to the command e.g `/bin/bash` or `cmd.exe /c`
-        # @option opts [Hash] :environment A hash of environment variables that will be injected into the command
-        # @option opts [IO] :stdin An IO object that will be used to redirect STDIN for the docker command
-        def execute(*command, options)
-          command.unshift(options[:interpreter]) if options[:interpreter]
-          # Build the `--env` parameters
-          envs = []
-          if options[:environment]
-            options[:environment].each { |env, val| envs.concat(['--env', "#{env}=#{val}"]) }
+        # @param command [string] The command to run
+        def execute(command)
+          env_vars = []
+          # CODEREVIEW: Is there a better way to handle `docker exec ... PT_foo` not working well?
```

Comment:
This definitely seems strange, and I think the weirdness comes from the shell class primarily handling getting the command in the right format for the shell, but the bash shell has some built-in assumptions for passing env vars that don't play well with the `docker exec` command. I'm not sure how to get around it without

1. Passing `options` and building the full command from `conn.execute` instead of the shell class, or
1. Having the shell class call a method on the connection class to format the command correctly, or
1. Having a separate 'docker' shell that builds docker-specific commands

Reply:
Mmk, I think part of the problem here is that I'm not sure how many user-provided transports are going to need to define their own environment handling or not. Without overengineering anything, I ended up with logic that says "If the connection define and environment variable handler then use it, otherwise use the default" which is the handler that our existing transports use.This seems to work well enough for now without the obscene parsing-env-vars-out-of-the-string, and we can adjust it based on peoples needs if we ever get to the point of writing transport plugins.